lib/bolt/target.rb

```diff
@@ -157,6 +157,7 @@ class Target
 
     # Satisfies the Puppet datatypes API
     def self.from_asserted_hash(hash)
+      Puppet.lookup(:bolt_inventory).deprecation
```

Comment:
I think you will need a `from_asserted_args` to capture what you want. This will only warn on `Target.new({'uri'=>'foo','options'=>{...}})` not `Target.new('foo', {...})`.

Reply:
I'm not sure I follow. Why would I add a method only to add a deprecation method to it? My misunderstanding is probably coming from the lack of documentation around Target.new(<uri>, <options>)..

Reply:
I think the only way to throw the message you want is to add the `from_asserted_args` class method. Otherwise you will never be able determine if a new target is instantiated with `Target.new('foo', {...})`. Try this plan:
```
  Target.new('foo')
  Target.new('bar', {'host-key-check' => 'false'})
  Target.new({'uri' => 'baz', 'options' => {'host-key-check' => 'false'}})
```
The one you want to add a deprecation warning for is the second one (and maybe the third one?) currently i think it will only warn on the second one.

Reply:
I *think* puppet will actually just call `.new` if `from_asserted_args` isn't defined https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/types/p_object_type.rb#L520 Puppet uses `from_asserted_hash` and `from_asserted_args` to create datatypes by passing arguments from your `Target.new` call to the datatypes implementation class instantiator. Target is a datatype and `target.rb` has it's implementation classes. `from_asserted_hash` is called when `Target.new({'uri' => 'foo', 'options' => { ... }})` is called (like Cas said), but if you pass the attributes as arguments instead of as a hash then theoretically `from_asserted_args` is called with the arguments you passed it. The part that's confusing is that this API isn't really documented (that I can find...), this is pretty much from https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/types/p_object_type.rb

Reply:
Confirmed that `from_asserted_args` isn't called if it's not defined, it just calls `.new` with the arguments. So 
```
plan test::foo() {
  return Target.new('localhost', { 'name' => 'name' })
}
```
with a V1 inventory just calls `initialize`. I think you just need to add the deprecation call there as well

Reply:
I think if the intent is for only `Target.new('foo')` to be valid it makes sense to add a deprecation warning for both the second and third options. Updated to add `from_asserted_args`.

Reply:
I dont know what to think of warning on the third one. I see `Target.new({...})` as being basically the whole point of the target (2) API. Specifically I think the *only* reason to use Target.new is to be able to set each part of the target you care about in the initialization.

Reply:
`Target.new({'uri' => 'foo'})` is still valid as long as it doesn't have the options hash. That will still work with a V2 inventory. `Target.new({'uri' => 'foo', 'options' => { ... }})` should raise the warning and `Target.new('uri', { ... })` should also raise the warning.

Reply:
Yeah that makes sense to me ^^. Whole sale warning on Target.new({..}) is bad, but if there is `uri`, and `options` keys that makes sense to warn on.

Reply:
Yeah, I don't think we should warn on the third one - that will still work with Inventory v2. The intention here is to make sure that plans are compatible with Inventory v2, so any target calls that behave the same between inventory versions 1 and 2 should not raise a warning.

Reply:
Updated to warn on `Target.new(uri, {})` and `Target.new('uri' => ..., 'options' => ...)`.

Reply:
@donoghuc Sorry I thought we might actually accept an `options` key in that hash, but confirmed we don't :) So yeah, `Target.new({ 'uri' => 'foo'})` should not warn, `Target.new({ 'uri' => 'foo', 'options' => { ... })` should.

Reply:
wait really? https://github.com/puppetlabs/bolt/blob/a67b8013dd9121276217a8ab61ef675957262800/lib/bolt/target.rb#L160 looks like options is fine?

Reply:
Sorry, I meant in Inventory 2! I thought we might still support `options` key in inventory 2, but confirmed we don't. that's why I deleted my earlier comment :P Either way seems like Tom implemented the right behavior.

Reply:
ahh, that makes sense! yeah good riddence `options`. :)